src/lib.rs

```diff
@@ -675,13 +930,191 @@ impl Contract {
         notify::notify_edit_proposal(id, proposal_author)
     }
 
+    #[payable]
+    pub fn edit_rfp(
+        &mut self,
+        id: RFPId,
+        body: VersionedRFPBody,
+        labels: HashSet<String>,
+    ) -> Promise {
+        near_sdk::log!("edit_rfp");
+        self.edit_rfp_internal(id, body.clone(), labels)
+    }
+
+    #[payable]
+    pub fn cancel_rfp(&mut self, id: RFPId, proposals_to_cancel: Vec<ProposalId>, proposals_to_unlink: Vec<ProposalId>) -> Promise {
+        near_sdk::log!("cancel_rfp");
+        let rfp: RFP =
+            self.rfps.get(id.into()).unwrap_or_else(|| panic!("RFP id {} not found", id)).into();
+        let mut body = rfp.snapshot.body.latest_version();
+        body.timeline = RFPTimelineStatus::Cancelled;
+
+        for proposal_id in proposals_to_cancel {
+            let proposal: Proposal = self.get_proposal(proposal_id).into();
+            let proposal_timeline = proposal.snapshot.body.latest_version().timeline;
+            let review_status = proposal_timeline.get_review_status().clone();
+            self.edit_proposal_timeline(proposal_id, TimelineStatus::Cancelled(review_status));
+        }
+
+        for proposal_id in proposals_to_unlink {
+            self.edit_proposal_linked_rfp(proposal_id, None);
+        }
+
+        self.edit_rfp_timeline(id, RFPTimelineStatus::Cancelled)
+    }
+
+    #[payable]
+    pub fn edit_rfp_timeline(&mut self, id: RFPId, timeline: RFPTimelineStatus) -> Promise {
+        near_sdk::log!("edit_rfp_timeline");
+        let rfp: RFP =
+            self.rfps.get(id.into()).unwrap_or_else(|| panic!("RFP id {} not found", id)).into();
+        let mut body = rfp.snapshot.body.latest_version();
+        body.timeline = timeline;
+
+        self.edit_rfp_internal(id, body.into(), rfp.snapshot.labels)
+    }
+
+    fn edit_rfp_internal(
+        &mut self,
+        id: RFPId,
+        body: VersionedRFPBody,
+        labels: HashSet<String>,
+    ) -> Promise {
+        let editor_id: AccountId = env::predecessor_account_id();
+        require!(
+            self.is_allowed_to_write_rfps(editor_id.clone()),
+            "The account is not allowed to edit RFPs"
+        );
+
+        let mut rfp: RFP =
+            self.rfps.get(id.into()).unwrap_or_else(|| panic!("RFP id {} not found", id)).into();
+
+        let rfp_body = body.clone().latest_version();
+
+        if rfp_body.timeline.is_proposal_selected() {
+            let mut has_approved_proposal = false;
+            for proposal_id in self.get_rfp_linked_proposals(id) {
+                let proposal: Proposal = self
+                    .proposals
+                    .get(proposal_id.into())
+                    .unwrap_or_else(|| panic!("Proposal id {} not found", proposal_id))
+                    .into();
+                if proposal.snapshot.body.latest_version().timeline.is_approved() {
+                    has_approved_proposal = true;
+                    break;
+                }
+            }
```

Comment:
Use iterator methods:

```suggestion
            let has_approved_proposal = self.get_rfp_linked_proposals(id)
                .into_iter()
                .filter_map(|proposal_id| self.proposals.get(proposal_id.into()))
                .any(|proposal|  proposal.snapshot.body.latest_version().timeline.is_approved());
```